contracts/OffsetHelper.sol

```diff
@@ -257,60 +282,46 @@ contract OffsetHelper is OffsetHelperStorage {
         // instantiate pool token (NCT or BCT)
         IToucanPoolToken PoolTokenImplementation = IToucanPoolToken(_fromToken);
 
-        // auto redeem NCT for TCO2; will transfer automatically picked TCO2 to this contract
-        PoolTokenImplementation.redeemAuto(_amount);
+        // auto redeem pool token for TCO2; will transfer automatically picked TCO2 to this contract
+        (tco2s, amounts) = PoolTokenImplementation.redeemAuto2(_amount);
+
+        // update balances
         balances[msg.sender][_fromToken] -= _amount;
+        for (uint256 index = 0; index < tco2s.length; index++) {
+            balances[msg.sender][tco2s[index]] += amounts[index];
+        }
 
-        // update TCO2 balances of the user to reflect the redeeming
-        tco2Balance[msg.sender] += _amount;
+        emit Redeemed(msg.sender, _fromToken, tco2s, amounts);
     }
 
-    // @param _amount the amount of TCO2 to retire
-    // @param _pool the pool that will be used to get scoredTCO2s
-    function autoRetire(uint256 _amount, address _pool) public {
-        require(isRedeemable(_pool), "Can't use this pool for scoredTCO2s.");
+    // @param _tco2s the addresses of the TCO2s to retire
+    // @param _amounts the amounts to retire from the matching TCO2
+    function autoRetire(address[] memory _tco2s, uint256[] memory _amounts)
+        public
+    {
+        require(_tco2s.length > 0, "You need to have at least one TCO2.");
+
         require(
-            tco2Balance[msg.sender] >= _amount,
-            "You don't have enough TCO2 in this contract."
+            _tco2s.length == _amounts.length,
+            "You need an equal number of addresses and amounts"
         );
 
-        // instantiate pool token (NCT or BCT)
-        IToucanPoolToken PoolTokenImplementation = IToucanPoolToken(_pool);
-
-        // I'm attempting to loop over all possible TCO2s that the contract could have
-        // see the contract's balance for said TCO2
-        // and retire until the whole amount has been retired
-        uint256 remainingAmount = _amount;
         uint256 i = 0;
-        address[] memory scoredTCO2s = PoolTokenImplementation.getScoredTCO2s();
-        uint256 scoredTCO2Len = scoredTCO2s.length;
-        while (remainingAmount > 0 && i < scoredTCO2Len) {
-            address tco2 = scoredTCO2s[i];
-            uint256 balance = IERC20(tco2).balanceOf(address(this));
-            if (balance < 1e15) continue;
-            uint256 amountToRetire = remainingAmount > balance
-                ? balance
-                : remainingAmount;
-            IToucanCarbonOffsets(tco2).retire(amountToRetire);
-            remainingAmount -= amountToRetire;
+        while (i < _tco2s.length) {
+            if (_amounts[i] < 1e15) continue;
```

Comment:
We can remove this check now.

Reply:
Will do, but can you please remind me why is that? I remember you telling me this once before, but I can't remember the reason.

Reply:
We removed the kilo limitation from the RetirementCertificates contract before releasing in the end.